src/flask_wtf/csrf.py

```diff
@@ -77,42 +77,112 @@ def validate_csrf(data, secret_key=None, time_limit=None, token_key=None):
-    )
-    time_limit = _get_config(time_limit, "WTF_CSRF_TIME_LIMIT", 3600, required=False)
+        if not data:
+            raise ValidationError("The CSRF token is missing.")
+
+        if field_name not in session:
+            raise ValidationError("The CSRF session token is missing.")
 
-    if not data:
-        raise ValidationError("The CSRF token is missing.")
+        s = URLSafeTimedSerializer(secret_key, salt="wtf-csrf-token")
 
-    if field_name not in session:
-        raise ValidationError("The CSRF session token is missing.")
+        try:
+            token = s.loads(data, max_age=time_limit)
+        except SignatureExpired as e:
+            raise ValidationError("The CSRF token has expired.") from e
+        except BadData as e:
+            raise ValidationError("The CSRF token is invalid.") from e
+
+        if not hmac.compare_digest(session[field_name], token):
+            raise ValidationError("The CSRF tokens do not match.")
+    except Exception as e:
+        logger.info("Falling back to legacy CSRF validation.")
+        token_key = 'csrf_token' if token_key is None else token_key
+        is_valid = legacy_validate_csrf(
+            data=data,
+            secret_key=secret_key,
+            time_limit=time_limit,
+            token_key=token_key
+        )
+        if is_valid is False:
+            raise e
 
-    s = URLSafeTimedSerializer(secret_key, salt="wtf-csrf-token")
+
+def legacy_validate_csrf(data, secret_key=None, time_limit=None,
+                         token_key='csrf_token', url_safe=False):
+    """Validates CSRF tokens signed by flask_wtf < 0.14.
+
+    Taken from https://github.com/benchling/flask-wtf/blob/318eea7be584e1c1116fc9d010bbbe95ff0fde55/flask_wtf/csrf.py#L66-L114
+    """
+    import time
+
+    def to_bytes(text):
+        """Transform string to bytes."""
+        if isinstance(text, str):
+            text = text.encode('utf-8')
+        return text
+
+    delimiter = '--' if url_safe else '##'
+    if not data or delimiter not in data:
+        return False
 
     try:
-        token = s.loads(data, max_age=time_limit)
-    except SignatureExpired as e:
-        raise ValidationError("The CSRF token has expired.") from e
-    except BadData as e:
-        raise ValidationError("The CSRF token is invalid.") from e
-
-    if not hmac.compare_digest(session[field_name], token):
-        raise ValidationError("The CSRF tokens do not match.")
+        expires, hmac_csrf = data.split(delimiter, 1)
+    except ValueError:
+        return False  # unpack error
+
+    if time_limit is None:
+        time_limit = current_app.config.get('WTF_CSRF_TIME_LIMIT', 3600)
+
+    if time_limit:
+        try:
+            expires = int(expires)
+        except ValueError:
+            return False
+
+        now = int(time.time())
+        if now > expires:
+            return False
+
+    if not secret_key:
+        secret_key = current_app.config.get(
```

Comment:
if neither of these options are set, we raise a different type of error here compared to before; is that ok? in the new `validate` above, we  look for
```
secret_key = _get_config(
            secret_key,
            "WTF_CSRF_SECRET_KEY",
            current_app.secret_key,
            message="A secret key is required to use CSRF.",
        )
```
which raises a RuntimeError if it's not found; here, this errors at the `hmac.new` below with a TypeError
```
>>> hmac.new(None, None, digestmod=hashlib.sha1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.8/hmac.py", line 153, in new
    return HMAC(key, msg, digestmod)
  File "/usr/lib/python3.8/hmac.py", line 48, in __init__
    raise TypeError("key: expected bytes or bytearray, but got %r" % type(key).__name__)
TypeError: key: expected bytes or bytearray, but got 'NoneType'
```

idek if its possible for both of these to be unset tho

Reply:
ooh, i actually cant tell you how secret_key gets set in our application -- but this `legacy_validate_csrf` is taken from the version of flask-wtf that we're currently running.. so somehow these values are all present when they need to be.